File: Retrospective/pipeline.py
#!/usr/bin/env python
# coding=utf-8

# ==============================================================================
# title : pipeline.py
# description : Import EEG files in MNE Python & extract alpha infos
# input: EDF+ files
# output: results summary in CSV file + figure for each participant
# author : G. Dumas, Human Genetics & Cognitive Functions, Institut Pasteur
# date : 2018
# version : 2.0
# usage : python pipeline.py
# python_version : 3.6
# license : BSD (3-clause)
# ==============================================================================

import os
import os.path as op
import mne
import numpy as np
import matplotlib
from copy import deepcopy
import matplotlib.pyplot as plt
from scipy.stats import kurtosis
from mne.time_frequency import psd_welch
from scipy.stats import ttest_ind

# Initialisation
matplotlib.use('Agg')
np.random.seed(42)
mne.set_log_level(verbose='CRITICAL')

# Parameters
subjectsNum = 135
inputFolder = 'raw'
outputFolder = 'results'

if not os.path.exists(outputFolder):
    os.mkdir(outputFolder)

# Header of the CSV file summarizing all the results
with open(os.path.join(outputFolder, 'Python_results.csv'), 'w') as outcsv:
    outcsv.write("Subject,Session,Group,BadChannels,EC,EO,AlphaChannel,AlphaFreq,AlphaPeak\n")

# Loop over groups
for subjectType in ['', 'T']:

    # Loop over participants
    for subjectNum in range(subjectsNum + 1):
        input_fname_mask = 'EEG-%s%03d' % \
                           (subjectType, subjectNum + 1)
        f_names = [f for f in os.listdir(inputFolder) if input_fname_mask in f]
        nsession = len(f_names)

        # Loop over sessions
        for f_name in f_names:
            print('\n', f_name)
            with open(os.path.join(outputFolder, 'Python_results.csv'), 'a') as outcsv:
                outcsv.write("{},{},{}".format(subjectNum + 1,
                                               int(f_name[f_name.find('_') + 1:-4]),
                                               subjectType))
            data = mne.io.read_raw_edf(op.join(inputFolder, f_name),
                                       montage=None, eog=None, misc=None,
                                       stim_channel=-1, annot=None,
                                       annotmap=None, preload=True,
                                       verbose=False)
            output_fname = deepcopy(f_name)

            # Handle channels
            for ich, lch in enumerate(data.info['ch_names']):
                newlabel = lch.replace('EEG ', '').replace('-Ref', '')
                data.info['ch_names'][ich] = newlabel
                data.info['chs'][ich]['ch_name'] = newlabel

            for ch in [u'POL E', u'POL PG1', u'POL PG2', u'POL T1',
                       u'POL T2', u'POL X1', u'POL X2', u'POL X3',
                       u'POL X4', u'POL X5', u'POL X6', u'POL X7']:
                try:
                    data.set_channel_types({ch: 'misc'})
                except:
                    print(ch, "is absent!")

            # Setup 10/20 montage
            montage = mne.channels.read_montage('standard_1020')
            data.set_montage(montage)

            # Extract the bad or missing channels
            pick_chans = [p for p in data.ch_names if
                          ('POL' not in p) and ('STI' not in p) and
                          ('A1' not in p) and ('A2' not in p) and
                          (p not in data.info['bads'])]
            data.drop_channels([p for p in data.ch_names
                                if ('POL' in p) or ('STI' in p) or
                                   ('A1' in p) or ('A2' in p) or
                                   (p in data.info['bads'])])

            # Filtering
            data.filter(phase='zero',
                        fir_window='hamming',
                        fir_design='firwin',
                        l_freq=None, h_freq=48,
                        l_trans_bandwidth='auto',
                        h_trans_bandwidth='auto',
                        filter_length='auto')

            # Plot SD & Kurtosis
            SD = data._data.std(axis=1)
            K = kurtosis(data._data.T)

            fname = op.join(outputFolder,
                            output_fname.replace('.edf', '_topo.pdf'))
            fig = plt.figure()
            plt.subplot(1, 2, 1)
            mne.viz.plot_topomap(SD, data.info, show=False,
                                 show_names=True, names=pick_chans)
            plt.title('Standard Deviation')
            plt.subplot(1, 2, 2)
            mne.viz.plot_topomap(K, data.info, show=False,
                                 show_names=True, names=pick_chans)
            plt.title('Kurtosis')
            plt.tight_layout()
            plt.savefig(fname)

            # Detect bad channels
            mask = SD > 0
            change = True
            while change:
                newmask = mask
                outlierSD = SD > (SD[mask].mean() + 2 * SD[mask].std())
                outlierK = K > (K[mask].mean() + 2 * K[mask].std())
                newmask[outlierSD] = False
                newmask[outlierK] = False
                if (newmask != mask).any():
                    mask[outlierSD] = False
                    mask[outlierK] = False
                    change = True
                else:
                    change = False
            data.info['bads'] += [ch for i, ch in enumerate(data.ch_names)
                                  if not mask[i]]

            # Consolidate the bad channels
            pick_chans = [p for p in data.ch_names if
                          ('POL' not in p) and ('STI' not in p) and
                          ('A1' not in p) and ('A2' not in p) and
                          (p not in data.info['bads'])]
            drop_chans = [p for p in data.ch_names
                          if ('POL' in p) or ('STI' in p) or
                             ('A1' in p) or ('A2' in p) or
                             (p in data.info['bads'])]
            print("Dropped channels:")
            for ch in drop_chans:
                print(ch)
            data.drop_channels(drop_chans)
            with open(os.path.join(outputFolder, 'Python_results.csv'), 'a') as outcsv:
                outcsv.write(',' + str(len(drop_chans)))

            # Extract events
            events = data._raw_extras[0]['events']
            startRec = 0
            stopRec = events[-1][0]
            unknowns = []

            # Detect eyes open and eyes closed intervals
            eyesOpened = 0
            eyesClosed = 0
            EO = []
            EC = []
            STOP = False
            for ievt, levt in enumerate(events):
                label = levt[2].lower()
                if label[0] != '+' and label[0] != '*':
                    unknown = True
                    if 'y' in label and not(STOP):
                        if 'o' in label:
                            # print("Eyes open at", levt[0])
                            if eyesClosed > 0 and levt[0] - eyesClosed > 1.:
                                EC.append([eyesClosed, levt[0]])
                            eyesOpened = levt[0]
                        if 'f' in label and not(STOP):
                            # print("Eyes close at", levt[0])
                            if eyesOpened > 0 and levt[0] - eyesOpened > 1.:
                                EO.append([eyesOpened, levt[0]])
                            eyesClosed = levt[0]
                        unknown = False

            # Perform spectral analyses
            fmin, fmax = 2, 48  # look at frequencies between 2 and 48Hz
            n_fft = 1024  # the FFT size (n_fft). Ideally a power of 2

            tmin, tmax = startRec, stopRec
            picks = [i for i, ch in enumerate(data.ch_names)
                     if 'o' in ch[0].lower() or 'p' in ch[0].lower()]
            channels = [ch for i, ch in enumerate(data.ch_names)
                        if 'o' in ch[0].lower() or 'p' in ch[0].lower()]

            psds, freqs = psd_welch(data, n_fft=1024, n_overlap=512,
                                    tmin=tmin, tmax=tmax,
                                    fmin=fmin, fmax=fmax,
                                    proj=False, picks=picks,
                                    n_jobs=1, verbose=False)

            psds = 10 * np.log10(psds)
            psds_mean = psds.mean(0)
            psds_std = psds.std(0)

            # Plot the global spectrum
            fname = op.join(outputFolder,
                            output_fname.replace('.edf', '_spectrum.pdf'))
            plt.figure()
            ax = plt.axes()
            ax.plot(freqs, psds_mean, color='k')
            ax.fill_between(freqs, psds_mean - psds_std,
                            psds_mean + psds_std,
                            color='k', alpha=.5)
            ax.set(title='Welch PSD', xlabel='Frequency',
                   ylabel='Power Spectral Density (dB)')
            plt.savefig(fname)

            def poweeg(data, tmin, tmax):
                """Extract Power Spectrum from EEG data and timing."""
                psds, freqs = psd_welch(data, n_fft=1024, n_overlap=512,
                                        tmin=tmin, tmax=tmax,
                                        fmin=fmin, fmax=fmax,
                                        proj=False, picks=picks,
                                        n_jobs=1, verbose=False)
                return psds, freqs

            # Compare the spectra for EO & EC
            EO_Global = np.zeros([len(EO), psds.shape[0], psds.shape[1]])
            goodEO = []
            for i, times in enumerate(EO):
                # print(i, times[0], times[1])
                try:
                    psds, freqs = poweeg(data, times[0], times[1])
                    EO_Global[i, :, :] = psds
                    goodEO.append(i)
                except:
                    print(times[1] - times[0])

            EC_Global = np.zeros([len(EC), psds.shape[0], psds.shape[1]])
            goodEC = []
            for i, times in enumerate(EC):
                # print(i, times[0], times[1])
                try:
                    psds, freqs = poweeg(data, times[0], times[1])
                    EC_Global[i, :, :] = psds
                    goodEC.append(i)
                except:
                    print(times[1] - times[0])

            print("EC length:", len(goodEC))
            print("EO length:", len(goodEO))
            with open(os.path.join(outputFolder, 'Python_results.csv'), 'a') as outcsv:
                outcsv.write(',' + str(len(goodEC)) +
                             ',' + str(len(goodEO)))

            [t, p] = ttest_ind(EC_Global[goodEC, :, :],
                               EO_Global[goodEO, :, :],
                               axis=0)

            # Search for the alpha peak
            mask = (freqs > 5) * (freqs < 20)
            try:
                tmp = t[:, mask]
                idx = np.argwhere(tmp == tmp.max())
                c_peak, f_peak, a_peak = channels[idx[0][0]], freqs[mask][idx[0][1]], tmp.max()
            except:
                c_peak, f_peak, a_peak = np.nan, np.nan, np.nan

            with open(os.path.join(outputFolder, 'Python_results.csv'), 'a') as outcsv:
                outcsv.write(',' + str(c_peak) +
                             ',' + str(f_peak) +
                             ',' + str(a_peak) +
                             '\n')

            # Plot EC vs. EO spectra for alpha peak detection
            fname = op.join(outputFolder,
                            output_fname.replace('.edf', '_EO-vs-EC.pdf'))

            if not(np.isnan(t).any()):
                plt.figure()
                plt.subplot(2, 1, 1)
                plt.plot(t.T)
                xt = range(0, len(freqs), 10)
                plt.xticks(xt, np.round(freqs[xt] * 10.) / 10.)
                plt.legend(channels)
                plt.title('EC vs. EO')
                plt.xlabel('Frequency')
                plt.ylabel('T-value')
                plt.subplot(2, 1, 2)
                plt.plot(-np.log(p.T))
                plt.xticks(xt, np.round(freqs[xt] * 10.) / 10.)
                plt.legend(channels)
                plt.xlabel('Frequency')
                plt.ylabel('-log(p-value)')
                index = np.argmax(np.max(-np.log(p)) == -np.log(p))
                ich, ifr = np.unravel_index(index, p.shape)
                plt.title('Maximum at ' + str(freqs[ifr]) +
                          'Hz for ' + channels[ich])
                plt.tight_layout()
                plt.savefig(fname)
            else:
                print("Error for contrast EC vs. EO")

File: Retrospective/pipeline.m
% ==============================================================================
% title : pipeline.m
% description : Import NihonKoden EEG files in EEGlab and calculate alpha infos
% input: EDF+ files
% output: results summary in CSV file + figure/log for each participant
% author : G. Dumas, Human Genetics & Cognitive Functions, Institut Pasteur
% date : 2018
% version : 2.0
% license : BSD (3-clause)
% ==============================================================================

%% Initialisation
rng(42)
addpath(genpath('PATH_TO_EEGLAB'));
eeglab('nogui');
dataFolder = pwd;

% Parameters
subjectsNum = 135;
inputFolder = 'raw';
outputFolder = 'results';
subjectGroups = {'', 'T'};

mkdir(outputFolder)
fileIDG = fopen([outputFolder,'/Matlab_results.csv'], 'a');

% Header of the CSV file summarizing all the results
fprintf(fileIDG,'Subject,Session,Group,EO,EC,BadChannels,AlphaType,AlphaFreq,AlphaPeak,AlphaChannel,ProcessingDuration\n');

for subjectGroup = 1:2
    %% loop on subjects
    for subjNum = 1:subjectsNum
        display(['Processing subject #', num2str(subjNum)])
        %% list all files
        subjFileBase = ['EEG-', subjectGroups{subjectGroup}, sprintf('%03d', subjNum)];
        files = dir([dataFolder, '/', inputFolder, '/', subjFileBase, '*_0001.edf']);

        if length(files)>1
            display('Multisession subject!')
        end

        if length(files)==0
            display('Missing subject!')
        end

        %% loop on sessions
        for f = 1:length(files)
            tic
            filename = files(f).name;
            display(['Session ', num2str(f)])
            fprintf(fileIDG,'%d,%d,%s,',subjNum,f, subjectGroups{subjectGroup});

            %% Individual log file
            fileID = fopen([outputFolder,'/',filename(1:end-4),'.log'], 'w');
            fprintf(fileID,[filename(1:end-4),'\n']);
            fprintf(fileID,[datestr(clock,'dd-mmm-yyyy HH:MM:SS'),'\n']);

            %% open files
            EEG = pop_biosig( [inputFolder,'/',filename] ,'importevent','off');
            EEG = eeg_checkset( EEG );
            [data,header] = lab_read_edf( [inputFolder,'/',filename] );
            timing = toc;
            fprintf(fileID,'Loading duration: %f seconds\n\n',timing);

            %% channels renamming
            EEG.chanlocs(1).labels = 'Fp1';
            EEG.chanlocs(2).labels = 'Fp2';
            EEG.chanlocs(3).labels = 'F3';
            EEG.chanlocs(4).labels = 'F4';
            EEG.chanlocs(5).labels = 'C3';
            EEG.chanlocs(6).labels = 'C4';
            EEG.chanlocs(7).labels = 'P3';
            EEG.chanlocs(8).labels = 'P4';
            EEG.chanlocs(9).labels = 'O1';
            EEG.chanlocs(10).labels = 'O2';
            EEG.chanlocs(11).labels = 'F7';
            EEG.chanlocs(12).labels = 'F8';
            EEG.chanlocs(13).labels = 'T3';
            EEG.chanlocs(14).labels = 'T4';
            EEG.chanlocs(15).labels = 'T5';
            EEG.chanlocs(16).labels = 'T6';
            EEG.chanlocs(17).labels = 'Fz';
            EEG.chanlocs(18).labels = 'Cz';
            EEG.chanlocs(19).labels = 'Pz';
            EEG = eeg_checkset( EEG );

            %% add channels positions
            EEG=pop_chanedit(EEG, 'lookup','standard-10-5-cap385.elp');
            EEG = eeg_checkset( EEG );

            %% Extract events
            for ch = 1:length(EEG.chanlocs)
                channels{ch} = EEG.chanlocs(ch).labels;
            end

            try
                eventsEDF = [EEG.event.latency];
            catch
                eventsEDF = [EEG.times(1),EEG.times(end)];
            end

            binEO=1;
            binEC=1;
            EyesOpen=[];
            EyesClose=[];

            accu=0;
            for n=1:length(header.events.TYP)
                lab = header.events.TYP{n};

                if strcmp(lab,'Yeux_ouverts')|strcmp(lab,'YEUX_OUVERTS')|strcmp(lab,'YO')
                    display([lab,'+@+',num2str(header.events.POS(n)),'+++',num2str(header.events.POS(n)-accu)])
                    EyesOpen(binEO)=header.events.POS(n);
                    binEO=binEO+1;
                    accu=header.events.POS(n);
                else
                    if (length(lab)>6 & strcmp(lab(1:7),'Yeux_fe'))|strcmp(lab,'Yeux_ferm?s')|strcmp(lab,'Yeux_ferm?s')|strcmp(lab,'YEUX_FERMES')|strcmp(lab,'YF')
                        display([lab,'-@-',num2str(header.events.POS(n)),'---',num2str(header.events.POS(n)-accu)])
                        EyesClose(binEC)=header.events.POS(n);
                        binEC=binEC+1;
                        accu=header.events.POS(n);
                    end
                end
            end

            fprintf(fileIDG,'%d,',length(EyesOpen));
            if length(EyesOpen)>0
                fprintf(fileID,'Eyes open: %d\n',length(EyesOpen));
            else
                fprintf(fileID,'Eyes open: none\n');
            end

            fprintf(fileIDG,'%d,',length(EyesClose));
            if length(EyesClose)>0
                fprintf(fileID,'Eyes close: %d\n',length(EyesOpen));
            else
                fprintf(fileID,'Eyes close: none\n');
            end

            %% Filtering
            EEG = pop_eegfiltnew(EEG, 1, 40, 1650, 0, [], 0);
            EEG = eeg_checkset( EEG );

            %% Detect bad channels and interpolate
            badChannels=zeros(1,19);done=0;
            while done==0
                newSTD=std(EEG.data(badChannels==0,:)');
                newBadChannels=(newSTD>=(mean(newSTD)+2*std(newSTD)));
                if sum(newBadChannels)==0
                    done=1;
                else
                    badChannels(badChannels==0)=newBadChannels;
                end
                badChannels;
            end
            display(channels(badChannels==1))

            EEG = pop_interp(EEG, find(badChannels), 'spherical'); %'spacetime');
            EEG = eeg_checkset( EEG );

            idxBad=find(badChannels);
            fprintf(fileIDG,'%d,',length(idxBad));
            if length(idxBad)>0
                fprintf(fileID,'Bad channels: %d (',length(idxBad));
                for bin = 1:length(idxBad)-1
                    fprintf(fileID,'%s, ',channels{idxBad(bin)});
                end
                if isempty(bin)
                    bin = 0;
                end
                fprintf(fileID,'%s)\n',channels{idxBad(bin+1)});
            else
                fprintf(fileID,'Bad channels: none\n')
            end

            %% Compute spectra
            [spectra,freqs] = spectopo(EEG.data(1:19,:),0,500,'freqrange',[1 40],'plot','off','freqfac',4);
            freqLabels=freqs(freqs<40);
            spectraFull=spectra(:,freqs<40);

            if length(EyesOpen)>0
                spectraEO=zeros(size(spectraFull));
                for binEO = 1:length(EyesOpen)
                    if EyesOpen(binEO)+4000<size(EEG.data,2)
                        if isempty(find(abs(sum(EEG.data(1:19,EyesOpen(binEO)+1000:EyesOpen(binEO)+4000),2))<0.0000001,1))
                            [spectra,freqs] = spectopo(EEG.data(1:19,EyesOpen(binEO)+1000:EyesOpen(binEO)+4000),0,500,'freqrange',[1 40],'plot','off','freqfac',4);
                            spectraEO=spectraEO+spectra(:,freqs<40);
                        else
                            binEO=binEO-1;
                        end
                    else
                        binEO=binEO-1;
                        break
                    end
                end
                spectraEO=spectraEO/binEO;
            else
                spectraEO=zeros(size(spectraFull));
            end

            if length(EyesClose)>0
                spectraEC=zeros(size(spectraFull));
                for binEC = 1:length(EyesClose)
                    if EyesClose(binEC)+4000<size(EEG.data,2)
                        if isempty(find(abs(sum(EEG.data(1:19,EyesClose(binEC)+1000:EyesClose(binEC)+4000),2))<0.0000001,1))
                            [spectra,freqs] = spectopo(EEG.data(1:19,EyesClose(binEC)+1000:EyesClose(binEC)+4000),0,500,'freqrange',[1 40],'plot','off','freqfac',4);
                            spectraEC=spectraEC+spectra(:,freqs<40);
                        else
                            binEO=binEO-1;
                        end
                    else
                        binEC=binEC-1;
                        break
                    end
                end
                spectraEC=spectraEC/binEC;
            else
                spectraEC=zeros(size(spectraFull));       
            end

            %% Plot spectra
            figure(998)
            clf
            imagesc(spectraFull)
            set(gca,'YTick',1:19,'YTickLabel',channels(1:19))
            set(gca,'XTick',1:20:160,'XTickLabel',freqLabels(1:20:160))
            for badChannel = find(badChannels)
                line([1 length(freqLabels)],badChannel*ones(1,2),'color',[1 0 0])
            end
            colorbar
            title('Spectra Full')
            print('-f998','-djpeg',[outputFolder,'/',filename(1:end-4),'_SpectraFull.jpg'])

            figure(999)
            clf
            if length(EyesClose)>0
                imagesc(spectraEC-spectraEO)
                set(gca,'YTick',1:19,'YTickLabel',channels(1:19))
                set(gca,'XTick',1:20:160,'XTickLabel',freqLabels(1:20:160))
                for badChannel = find(badChannels)
                    line([1 length(freqLabels)],badChannel*ones(1,2),'color',[1 0 0])
                end
                colorbar

                % Search alpha peak between 5Hz and 20Hz across posterior sensors
                fSearch=find((freqLabels>5)&(freqLabels<20));
                channelSearch=[7:10 19];
                tmp=spectraEC(channelSearch,fSearch)-spectraEO(channelSearch,fSearch);
                [I,J] = ind2sub(size(tmp),find(tmp==max(tmp(:))));
                alphaChannel=channels{channelSearch(I)};
                alphaPeak=freqLabels(fSearch(J));

                if length(EyesOpen)>0
                    title(sprintf('EC vs EO - Alpha peak at %fHz on channel %s', alphaPeak, alphaChannel))
                    fprintf(fileID,'Alpha peak (EC vs EO) at %fHz on channel %s\n', alphaPeak, alphaChannel);
                    fprintf(fileIDG,'ECvsEO,%f,%f,%s,',alphaPeak, max(tmp(:)), alphaChannel);
                else
                    title(sprintf('EC spectra - Alpha peak at %fHz on channel %s', alphaPeak, alphaChannel))
                    fprintf(fileID,'Alpha peak (EC) at %fHz on channel %s\n', alphaPeak, alphaChannel);
                    fprintf(fileIDG,'EC,%f,%f,%s,',alphaPeak, max(tmp(:)), alphaChannel);
                end
                print('-f999','-djpeg',[outputFolder,'/',filename(1:end-4),'_EC-vs-EO.jpg'])
            else
                fprintf(fileID,'No alpha peak detected...\n');
                fprintf(fileIDG,'None,NaN,NaN,NaN,');
            end

            timing2 = toc;

            %% Clean Workspace
            fprintf(fileID,'\n\nAnalysis duration: %f seconds\n',timing+timing2);
            fprintf(fileIDG,'%f\n',timing+timing2);
            save([outputFolder,'/',filename(1:end-4)], 'EyesClose', 'EyesOpen', 'badChannels', 'spectraEC', 'spectraEO', 'spectraFull', 'timing') 
            fclose(fileID)
            STUDY = []; CURRENTSTUDY = 0; ALLEEG = []; EEG=[]; CURRENTSET=[];
            clc;close all;
        end
    end
end
fclose(fileIDG);